metaflow/plugins/cards/card_datastore.py

```diff
@@ -0,0 +1,200 @@
+""" 
+
+"""
+
+from collections import namedtuple
+from hashlib import sha1
+from io import BytesIO
+import os
+import shutil
+
+from metaflow.datastore.local_storage import LocalStorage
+from metaflow.metaflow_config import (
+    DATASTORE_CARD_S3ROOT,
+    DATASTORE_CARD_LOCALROOT,
+    DATASTORE_LOCAL_DIR,
+    DATASTORE_CARD_SUFFIX,
+)
+
+from .exception import CardNotPresentException
+
+TEMP_DIR_NAME = "metaflow_card_cache"
+NUM_SHORT_HASH_CHARS = 5
+
+CardInfo = namedtuple("CardInfo", ["type", "hash"])
+
+
+def path_spec_resolver(pathspec):
+    run_id, step_name, task_id = None, None, None
+    splits = pathspec.split("/")
+    if len(splits) == 1:  # only flowname mentioned
+        return splits[0], run_id, step_name, task_id
+    elif len(splits) == 2:  # flowname , runid mentioned
+        return splits[0], splits[1], step_name, task_id
+    elif len(splits) == 3:  # flowname , runid , stepname
+        return splits[0], splits[1], splits[2], task_id
+    elif len(splits) == 4:  # flowname ,runid ,stepname , taskid
+        return splits[0], splits[1], splits[2], splits[3]
+
+
+class CardDatastore(object):
+    @classmethod
+    def get_storage_root(cls, storage_type):
+        if storage_type == "s3":
+            return DATASTORE_CARD_S3ROOT
+        else:
+            # Borrowing some of the logic from LocalStorage.get_storage_root
+            result = DATASTORE_CARD_LOCALROOT
+            if result is None:
+                current_path = os.getcwd()
+                check_dir = os.path.join(
+                    current_path, DATASTORE_LOCAL_DIR, DATASTORE_CARD_SUFFIX
+                )
+                check_dir = os.path.realpath(check_dir)
+                orig_path = check_dir
+                while not os.path.isdir(check_dir):
+                    print(check_dir)
+                    new_path = os.path.dirname(current_path)
+                    if new_path == current_path:
+                        break  # We are no longer making upward progress
+                    current_path = new_path
+                    check_dir = os.path.join(
+                        current_path, DATASTORE_LOCAL_DIR, DATASTORE_CARD_SUFFIX
+                    )
+                result = orig_path
+
+            return result
+
+    def __init__(self, flow_datastore, run_id, step_name, task_id, path_spec=None):
```

Comment:
nit: inconsistent path_spec vs pathspec. And why do we need both the run_id, step_name, task_id and pathspec. Isn't the info duplicated?

Reply:
Fixing this.